### PR TITLE
DIS positivity

### PR DESCRIPTION
**Pull request by felixhekhorn**
to be closed after https://github.com/N3PDF/yadism/pull/146 has been merged and yadism released

### CODE REVIEW

**Comment by felixhekhorn** [comment]
Comparison new vs old:

<details>
<summary>F2D</summary>
<pre>
$ pineappl diff results/NNPDF_POS_F2D_40-20220817132453/NNPDF_POS_F2D_40.pineappl.lz4 FK_POSF2DW.pineappl.lz4 NNPDF40_nnlo_as_01180 --ignore-orders --ignore-lumis --ignore-bin-limits
LHAPDF 6.4.0 loading /home/felix/local/share/LHAPDF/NNPDF40_nnlo_as_01180/NNPDF40_nnlo_as_01180_0000.dat
NNPDF40_nnlo_as_01180 PDF set, member #0, version 1; LHAPDF ID = 331100
b  x1                         x2                                        diff               
--+-+-+------------------------+------------------------+------------+------------+--------
 0 5 5                0.0000005                0.0000005 2.8158080e-1 2.8141248e-1 5.981e-4
 1 5 5 0.0000019407667236782136 0.0000019407667236782136 2.3615406e-1 2.3606068e-1 3.956e-4
 2 5 5  0.000007533150951473337  0.000007533150951473337 1.9745225e-1 1.9738759e-1 3.276e-4
 3 5 5  0.000029240177382128657  0.000029240177382128657 1.6479937e-1 1.6470902e-1 5.485e-4
 4 5 5   0.00011349672651536727   0.00011349672651536727 1.3748436e-1 1.3739885e-1 6.224e-4
 5 5 5   0.00044054134013486355   0.00044054134013486355 1.1456282e-1 1.1450772e-1 4.812e-4
 6 5 5    0.0017099759466766963    0.0017099759466766963 9.4558924e-2 9.4511973e-2 4.968e-4
 7 5 5     0.006637328831200572     0.006637328831200572 7.7957365e-2 7.7884518e-2 9.353e-4
 8 5 5      0.02576301385940815      0.02576301385940815 6.8709373e-2 6.8661997e-2 6.900e-4
 9 5 5                      0.1                      0.1 5.7479711e-2 5.7388453e-2 1.590e-3
10 5 5                     0.18                     0.18 4.7436192e-2 4.7308119e-2 2.707e-3
11 5 5                     0.26                     0.26 3.6689447e-2 3.6547517e-2 3.883e-3
12 5 5      0.33999999999999997      0.33999999999999997 2.6501562e-2 2.6336839e-2 6.254e-3
13 5 5      0.42000000000000004      0.42000000000000004 1.7973429e-2 1.7844343e-2 7.234e-3
14 5 5                      0.5                      0.5 1.0958372e-2 1.0879419e-2 7.257e-3
15 5 5                     0.58                     0.58 5.5489416e-3 5.4931997e-3 1.015e-2
16 5 5                     0.66                     0.66 2.1684957e-3 2.1435694e-3 1.163e-2
17 5 5                     0.74                     0.74 5.9505464e-4 5.7545637e-4 3.406e-2
18 5 5                     0.82                     0.82 1.3932600e-4 1.3566888e-4 2.696e-2
19 5 5                      0.9                      0.9 5.4007513e-5 5.0210248e-5 7.563e-2
</pre>
</details>

<details>
<summary>F2U</summary>
<pre>
$ pineappl diff results/NNPDF_POS_F2U_40-20220817132509/NNPDF_POS_F2U_40.pineappl.lz4 FK_POSF2U.pineappl.lz4 NNPDF40_nnlo_as_01180 --ignore-orders --ignore-lumis --ignore-bin-limits
LHAPDF 6.4.0 loading /home/felix/local/share/LHAPDF/NNPDF40_nnlo_as_01180/NNPDF40_nnlo_as_01180_0000.dat
NNPDF40_nnlo_as_01180 PDF set, member #0, version 1; LHAPDF ID = 331100
b  x1                         x2                                        diff               
--+-+-+------------------------+------------------------+------------+------------+--------
 0 5 5                0.0000005                0.0000005  1.1274732e0  1.1268024e0 5.953e-4
 1 5 5 0.0000019407667236782136 0.0000019407667236782136 9.4617628e-1 9.4580687e-1 3.906e-4
 2 5 5  0.000007533150951473337  0.000007533150951473337 7.9196039e-1 7.9170818e-1 3.186e-4
 3 5 5  0.000029240177382128657  0.000029240177382128657 6.6221478e-1 6.6186385e-1 5.302e-4
 4 5 5   0.00011349672651536727   0.00011349672651536727 5.5424982e-1 5.5392592e-1 5.847e-4
 5 5 5   0.00044054134013486355   0.00044054134013486355 4.6456376e-1 4.6437428e-1 4.080e-4
 6 5 5    0.0017099759466766963    0.0017099759466766963 3.8790331e-1 3.8776300e-1 3.618e-4
 7 5 5     0.006637328831200572     0.006637328831200572 3.2804462e-1 3.2780191e-1 7.404e-4
 8 5 5      0.02576301385940815      0.02576301385940815 3.0545624e-1 3.0528532e-1 5.599e-4
 9 5 5                      0.1                      0.1 3.0186266e-1 3.0173775e-1 4.140e-4
10 5 5                     0.18                     0.18 2.9849008e-1 2.9818749e-1 1.015e-3
11 5 5                     0.26                     0.26 2.7827298e-1 2.7757600e-1 2.511e-3
12 5 5      0.33999999999999997      0.33999999999999997 2.4164915e-1 2.4051524e-1 4.715e-3
13 5 5      0.42000000000000004      0.42000000000000004 1.9460850e-1 1.9362557e-1 5.076e-3
14 5 5                      0.5                      0.5 1.4250671e-1 1.4153918e-1 6.836e-3
15 5 5                     0.58                     0.58 9.4278593e-2 9.3400529e-2 9.401e-3
16 5 5                     0.66                     0.66 5.5287774e-2 5.4767371e-2 9.502e-3
17 5 5                     0.74                     0.74 2.6433603e-2 2.6160118e-2 1.045e-2
18 5 5                     0.82                     0.82 8.6562824e-3 8.4611946e-3 2.306e-2
19 5 5                      0.9                      0.9 1.3254605e-3 1.2610145e-3 5.111e-2
</pre>
</details>

<details>
<summary>F2S</summary>
<pre>
$ pineappl diff results/NNPDF_POS_F2S_40-20220817132526/NNPDF_POS_F2S_40.pineappl.lz4 FK_POSF2S.pineappl.lz4 NNPDF40_nnlo_as_01180 --ignore-orders --ignore-lumis --ignore-bin-limits
LHAPDF 6.4.0 loading /home/felix/local/share/LHAPDF/NNPDF40_nnlo_as_01180/NNPDF40_nnlo_as_01180_0000.dat
NNPDF40_nnlo_as_01180 PDF set, member #0, version 1; LHAPDF ID = 331100
b  x1                         x2                                        diff                
--+-+-+------------------------+------------------------+------------+------------+---------
 0 5 5                0.0000005                0.0000005 2.8021356e-1 2.8006298e-1  5.377e-4
 1 5 5 0.0000019407667236782136 0.0000019407667236782136 2.3422194e-1 2.3412424e-1  4.173e-4
 2 5 5  0.000007533150951473337  0.000007533150951473337 1.9468476e-1 1.9459611e-1  4.555e-4
 3 5 5  0.000029240177382128657  0.000029240177382128657 1.6077240e-1 1.6064410e-1  7.986e-4
 4 5 5   0.00011349672651536727   0.00011349672651536727 1.3151968e-1 1.3138004e-1  1.063e-3
 5 5 5   0.00044054134013486355   0.00044054134013486355 1.0556581e-1 1.0544314e-1  1.163e-3
 6 5 5    0.0017099759466766963    0.0017099759466766963 8.0776629e-2 8.0664849e-2  1.386e-3
 7 5 5     0.006637328831200572     0.006637328831200572 5.6287445e-2 5.6187510e-2  1.779e-3
 8 5 5      0.02576301385940815      0.02576301385940815 3.3511842e-2 3.3452870e-2  1.763e-3
 9 5 5                      0.1                      0.1 1.2490001e-2 1.2439917e-2  4.026e-3
10 5 5                     0.18                     0.18 7.0452774e-3 7.0042038e-3  5.864e-3
11 5 5                     0.26                     0.26 4.0652664e-3 4.0451726e-3  4.967e-3
12 5 5      0.33999999999999997      0.33999999999999997 2.0578146e-3 2.0517170e-3  2.972e-3
13 5 5      0.42000000000000004      0.42000000000000004 1.3361661e-3 1.3241663e-3  9.062e-3
14 5 5                      0.5                      0.5 1.4380389e-3 1.4281362e-3  6.934e-3
15 5 5                     0.58                     0.58 1.3822946e-3 1.3752288e-3  5.138e-3
16 5 5                     0.66                     0.66 9.1888491e-4 9.1378082e-4  5.586e-3
17 5 5                     0.74                     0.74 4.0045997e-4 3.9526491e-4  1.314e-2
18 5 5                     0.82                     0.82 1.1951886e-4 1.1563386e-4  3.360e-2
19 5 5                      0.9                      0.9 3.5331686e-5 3.5886541e-5 -1.546e-2
</pre>
</details>

<details>
<summary>F2C</summary>
<pre>
$ pineappl diff results/NNPDF_POS_F2C_40-20220817132541/NNPDF_POS_F2C_40.pineappl.lz4 FK_POSF2C.pineappl.lz4 NNPDF40_nnlo_as_01180 --ignore-orders --ignore-lumis --ignore-bin-limits
LHAPDF 6.4.0 loading /home/felix/local/share/LHAPDF/NNPDF40_nnlo_as_01180/NNPDF40_nnlo_as_01180_0000.dat
NNPDF40_nnlo_as_01180 PDF set, member #0, version 1; LHAPDF ID = 331100
b  x1                         x2                                        diff                
--+-+-+------------------------+------------------------+------------+------------+---------
 0 5 5                0.0000005                0.0000005 7.3109984e-1 7.3656144e-1 -7.415e-3
 1 5 5 0.0000019407667236782136 0.0000019407667236782136 5.8512229e-1 5.8890537e-1 -6.424e-3
 2 5 5  0.000007533150951473337  0.000007533150951473337 4.5508294e-1 4.5772159e-1 -5.765e-3
 3 5 5  0.000029240177382128657  0.000029240177382128657 3.4026157e-1 3.4197415e-1 -5.008e-3
 4 5 5   0.00011349672651536727   0.00011349672651536727 2.3985446e-1 2.4082438e-1 -4.028e-3
 5 5 5   0.00044054134013486355   0.00044054134013486355 1.5349946e-1 1.5398691e-1 -3.166e-3
 6 5 5    0.0017099759466766963    0.0017099759466766963 8.2213212e-2 8.2565064e-2 -4.262e-3
 7 5 5     0.006637328831200572     0.006637328831200572 3.1102813e-2 3.1163557e-2 -1.949e-3
 8 5 5      0.02576301385940815      0.02576301385940815 4.1051058e-3 4.1525658e-3 -1.143e-2
 9 5 5                      0.1                      0.1 2.1393980e-3 2.1369423e-3  1.149e-3
10 5 5                     0.18                     0.18 6.4430293e-3 6.4151886e-3  4.340e-3
11 5 5                     0.26                     0.26 7.7173902e-3 7.7241680e-3 -8.775e-4
12 5 5      0.33999999999999997      0.33999999999999997 7.5264177e-3 7.5624729e-3 -4.768e-3
13 5 5      0.42000000000000004      0.42000000000000004 5.3661665e-3 5.3313968e-3  6.522e-3
14 5 5                      0.5                      0.5 2.5933246e-3 2.5590248e-3  1.340e-2
15 5 5                     0.58                     0.58 9.0758630e-4 9.2021892e-4 -1.373e-2
16 5 5                     0.66                     0.66 3.1576829e-4 3.3997595e-4 -7.120e-2
17 5 5                     0.74                     0.74 1.7340076e-4 1.7631502e-4 -1.653e-2
18 5 5                     0.82                     0.82 7.5874933e-5 7.0327655e-5  7.888e-2
19 5 5                      0.9                      0.9 1.5290310e-5 1.1812942e-5  2.944e-1
</pre>
</details>

<details>
<summary>F2LL</summary>
<pre>
$ pineappl diff results/NNPDF_POS_FLL_40-20220817132618/NNPDF_POS_FLL_40.pineappl.lz4 FK_POSFLL.pineappl.lz4 NNPDF40_nnlo_as_01180 --ignore-orders --ignore-lumis --ignore-bin-limits
LHAPDF 6.4.0 loading /home/felix/local/share/LHAPDF/NNPDF40_nnlo_as_01180/NNPDF40_nnlo_as_01180_0000.dat
NNPDF40_nnlo_as_01180 PDF set, member #0, version 1; LHAPDF ID = 331100
b  x1                         x2                                        diff                
--+-+-+------------------------+------------------------+------------+------------+---------
 0 5 5                0.0000005                0.0000005 2.2460574e-2 2.4099144e-2 -6.799e-2
 1 5 5 0.0000019407667236782136 0.0000019407667236782136 3.8903511e-2 4.0305031e-2 -3.477e-2
 2 5 5  0.000007533150951473337  0.000007533150951473337 5.2163800e-2 5.3361314e-2 -2.244e-2
 3 5 5  0.000029240177382128657  0.000029240177382128657 6.2648490e-2 6.3673031e-2 -1.609e-2
 4 5 5   0.00011349672651536727   0.00011349672651536727 7.1542764e-2 7.2414682e-2 -1.204e-2
 5 5 5   0.00044054134013486355   0.00044054134013486355 8.1394744e-2 8.2129108e-2 -8.942e-3
 6 5 5    0.0017099759466766963    0.0017099759466766963 9.3150683e-2 9.3771212e-2 -6.617e-3
 7 5 5     0.006637328831200572     0.006637328831200572 9.7637841e-2 9.8177603e-2 -5.498e-3
 8 5 5      0.02576301385940815      0.02576301385940815 8.2772497e-2 8.3218892e-2 -5.364e-3
 9 5 5                      0.1                      0.1 4.9128890e-2 4.9309187e-2 -3.656e-3
10 5 5                     0.18                     0.18 3.3740982e-2 3.3763170e-2 -6.571e-4
11 5 5                     0.26                     0.26 2.3816580e-2 2.3760779e-2  2.348e-3
12 5 5      0.33999999999999997      0.33999999999999997 1.5747872e-2 1.5666488e-2  5.195e-3
13 5 5      0.42000000000000004      0.42000000000000004 9.4561224e-3 9.3791572e-3  8.206e-3
14 5 5                      0.5                      0.5 5.0911219e-3 5.0383470e-3  1.047e-2
15 5 5                     0.58                     0.58 2.4156149e-3 2.3850165e-3  1.283e-2
16 5 5                     0.66                     0.66 9.6104156e-4 9.4713618e-4  1.468e-2
17 5 5                     0.74                     0.74 2.8938708e-4 2.8440804e-4  1.751e-2
18 5 5                     0.82                     0.82 5.5019575e-5 5.4532111e-5  8.939e-3
19 5 5                      0.9                      0.9 4.1290593e-6 4.8769048e-6 -1.533e-1
</pre>
</details>

**Comment by felixhekhorn** [comment]
The issue of TMC (discussed in https://github.com/N3PDF/yadism/pull/146 and various PC) is solved here by distinguishing `NCPositivityCharge: null`, which means "this is not a positivity dataset" and `NCPositivityCharge: all` which means "this is a positivity datset, but *without a restriction on the coupling quark*" - this is needed for FL_light (which should probe the gluon, I guess) and F2C (which comes with quark restriction *built-in*)

in effect this means:
- here we use: https://github.com/NNPDF/runcards/blob/04572c4124fb1a379ef77805749d726f0cd6a3e9/pinefarm/external/yad.py#L38-L40
- instead in yadism we use: https://github.com/N3PDF/yadism/blob/3b7501fde12f29ff258fd1946b4848d0fc4fe78b/src/yadism/coefficient_functions/coupling_constants.py#L229-L237

File: pinefarm/external/yad.py [hunk 1]
```diff
@@ -35,6 +35,10 @@ def __init__(self, *args, **kwargs):
         ) as o:
             self.obs = yaml.safe_load(o)
 
+        # deactivate TMC for positivity observables
+        if self.obs["NCPositivityCharge"] is not None:
+            self.theory["TMC"] = 0
```
Comment (AleCandido) [inline pinefarm/external/yad.py:40]: If possible, add a reference to minutes :) (where we discussed why we want to use obs w/o TMC).

**Review by AleCandido — COMMENTED** [review]
Fine, the moment https://github.com/N3PDF/yadism/pull/146 is ready, also this one is good to go.

I would only regenerate the data, just in case there are updated descriptions or other metadata. I do not expect anything else :)

File: pinefarm/external/yad.py (hunk 1 above)
Comment (felixhekhorn) [reply pinefarm/external/yad.py:41]: done in https://github.com/NNPDF/runcards/pull/143/commits/bf2f20cba8d307d034fb9bceeedae4829d519b50

**Review by AleCandido — APPROVED** [review]

**Pull request merged**